client/trino-cli/src/test/java/io/trino/cli/TestClientRedirect.java

```diff
@@ -0,0 +1,316 @@
+            throws Exception
+    {
+        proxyServer.enqueue(new MockResponse()
+                .setResponseCode(307)
+                .addHeader(LOCATION, server.url("redirect/v1/statement"))
+                .addHeader(AUTHORIZATION, isBasicAuth ? DUMMY_BASIC_AUTH_TOKEN : DUMMY_ACCESS_TOKEN));
+        server.enqueue(new MockResponse()
+                .addHeader(CONTENT_TYPE, "application/json")
+                .setBody(createResults(server)));
+
+        QueryRunner queryRunner;
+        if (isBasicAuth) {
+            queryRunner = createQueryRunner(createTrinoUriBasicAuth(proxyServer, disableFollowRedirects), createClientSession(proxyServer));
+        }
+        else {
+            queryRunner = createQueryRunner(createTrinoUriAccessToken(proxyServer, disableFollowRedirects), createClientSession(proxyServer));
+        }
+
+        try (Query query = queryRunner.startQuery("SELECT 1")) {
+            query.renderOutput(getTerminal(), nullPrintStream(), nullPrintStream(), CSV, Optional.of(""), false);
+        }
+
+        //verify the redirect request has the same Authorization header
+        RecordedRequest redirectedRequest = server.takeRequest();
+        if (isBasicAuth) {
+            assertThat(redirectedRequest.getHeader(AUTHORIZATION)).isEqualTo(DUMMY_BASIC_AUTH_TOKEN);
+        }
+        else {
+            assertThat(redirectedRequest.getHeader(AUTHORIZATION)).isEqualTo("Bearer " + DUMMY_ACCESS_TOKEN);
+        }
+    }
+
+    private void assertRedirectIsNotFollowed(boolean isBasicAuth)
+    {
+        proxyServer.enqueue(new MockResponse()
+                .setResponseCode(307)
+                .addHeader(LOCATION, server.url("redirect/v1/statement"))
+                .addHeader(AUTHORIZATION, isBasicAuth ? DUMMY_BASIC_AUTH_TOKEN : DUMMY_ACCESS_TOKEN));
+        server.enqueue(new MockResponse()
+                .addHeader(CONTENT_TYPE, "application/json")
+                .setBody(createResults(server)));
+        try {
+            QueryRunner queryRunner;
+            if (isBasicAuth) {
+                queryRunner = createQueryRunner(createTrinoUriBasicAuth(proxyServer, Optional.of(true)), createClientSession(proxyServer));
+            }
+            else {
+                queryRunner = createQueryRunner(createTrinoUriAccessToken(proxyServer, Optional.of(true)), createClientSession(proxyServer));
+            }
+
+            try (Query query = queryRunner.startQuery("SELECT 1")) {
+                query.renderOutput(getTerminal(), nullPrintStream(), nullPrintStream(), CSV, Optional.of(""), false);
+            }
+        }
+        catch (Exception e) {
+            assertThat(e).isInstanceOf(ClientException.class);
+            assertThat(e.getMessage()).isEqualTo(REDIRECT_EXCEPTION);
+        }
+    }
+
+    static TrinoUri createTrinoUriAccessToken(MockWebServer server, Optional<Boolean> disableFollowRedirects)
```

Comment:
`createTrinoUriAccessToken`, `createTrinoUriBasicAuth` and `createTrinoUriExternalAuth` are mostly identical.
We should extract a helper function.